docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,318 @@
+
+For each type, we will synthesize one `public` constructor:
+
+The signature for the synthesized initializer for type `T` is:
+```csharp
+public T.__init(member0: typeof(member0) = default(member0), member1 : typeof(member1) = default(member1), ...)
+```
+where `(member0, member1, ... memberN)` is the set of members of `public` visibility, and `default(member0)`
+is the value defined by the initialization expression in `member0` if it exist, or the default value of `member0`'s type.
+If `member0`'s type is not default initializable and the the member doesn't provide an initial value, then the parameter will not have a default value.
+
+The body of the constructor will initialize each member with the value comming from the corresponding constructor argument if such argument exists,
+otherwise the member will be initialized to its default value either defined by the init expr of the member, or the default value of the type if the 
+type is default-initializable. If the member type is not default-initializable and a default value isn't provided on the member, then such the constructor
+synthesis will fail and the constructor will not be added to the type. Failure to synthesis a constructor is not an error, and an error will appear
+if the user is trying to initialize a value of the type in question assuming such a constructor exist.
+
+Note that if every member of a struct contains a default expression, the synthesized `__init` method can be called with 0 arguments, however, this will not cause a variable declaration to be implicitly initialized. Implicit initialization is a backward compatibility feature that only work for user-defined `__init()` methods.
+
+### Single argument constructor call
+
+Call to a constructor with a single argument is always treated as a syntactic sugar of type cast:
+```csharp
+int x = int(1.0f); // is treated as (int) 1.0f;
+MyType y = MyType(arg); // is treated as (MyType)arg;
+MyType x = MyType(y); // equivalent to `x = y`.
+```
+
+The compiler will attempt to resolve all type casts using type coercion rules, if that failed, will fall back to resolve it as a constructor call.
+
+### Initialization List
+
+Slang allows initialization of a variable by assigning it with an initialization list. 
+Generally, Slang will always try to resolve initialization list coercion as if it is an explicit constructor invocation.
+For example, given:
+```csharp
+S obj = {1,2};
+```
+Slang will try to convert the code into:
+```csharp
+S obj = S(1,2);
+```
+
+Following the same logic, an empty initializer list will translate into a default-initialization:
+```csharp
+S obj = {};
+// equivalent to:
+S obj = S();
+```
+
+Note that initializer list of a single argument still translates directly into a constructor call and not a type cast. For example:
+```csharp
+void test()
+{
+  MyType t = {1};
+  // translates to:
+  // MyType t = MyType.__init(1);
+  // which is not
+  // MyType t = MyType(t)
```

Comment:
I'm not quite understand this examples, what's the different between
`MyType t = MyType.__init(1);`
and
`MyType t = MyType(t)`

Reply:
MyType.__init(1) calls the ctor directly.

MyType(1) means (MyType)1 which will first try a set of builtin coercion rules to convert 1 to MyType. If the builtin rules doesn't apply, then fallback to calling MyType.__init(1).

Reply:
Oh, I see that we want to differentiate from
`Single argument constructor call` sub-section above?

Reply:
Yes. 

MyType s = {1}
Always calls the ctor.

MyType(1) goes through type cast.